stdlib/either.mli

```diff
@@ -0,0 +1,88 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*         Gabriel Scherer, projet Parsifal, INRIA Saclay                 *)
+(*                                                                        *)
+(*   Copyright 2019 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Either type.
+
+    Either is the simplest and most generic sum type. It is isomorphic
+    to the [result] type; it should be used when there is no implied
+    success/failure semantics in the choice of one of the constructors.
+
+    @since 4.11 *)
+
+(* Unlike [result], no [either] type is made available in Stdlib,
+   one needs to access [Either.t] explicitly:
+
+   - This type is less common in typical OCaml codebases,
+     which prefer domain-specific variant types whose constructors
+     carry more meaning.
+   - Adding this to Stdlib would raise warnings in existing codebases
+     that already use a constructor named Left or Right:
+     + when opening a module that exports such a name,
+       warning 45 is raised
+     + adding a second constructor of the same name in scope kicks
+       in the disambiguation mechanisms, and warning 41 may now
+       be raised by existing code.
+
+   If the use becomes more common in the future we can always
+   revisit this choice.
+*)
+
+type ('a, 'b) t = Left of 'a | Right of 'b (**)
+(** A value of [('a, 'b) Either.t] contains either a value of ['a]  or a value
+    of ['b] *)
+
+val left : 'a -> ('a, 'b) t
+(** [left v] is [Left v]. *)
+
+val right : 'b -> ('a, 'b) t
+(** [right v] is [Right v]. *)
+
+val find_left : ('a, 'b) t -> 'a option
+(** [find_left (Left v)] is [Some v], [find_left (Right _)] is [None] *)
```

Comment:
I may be alone in this, but I would prefer a function that provides a default for the case where the argument is `Right` rather than returning an option:
```
val from_left : ('a, 'b) t -> 'a -> 'a
```

Reply:
I think `Either.from_left` doesn't read right :-). Here again `Result` and `Option` should guide us (see my now outdated comment for a suggestion).

Reply:
If we let `Result` and `Option` guide us then we should add something like the `value` function in those modules, which is precisely what `from_left` is.  (Furthermore, Haskell has `fromLeft`, too.)

Reply:
Yes I suggested in the outdated comment (as per `{Result,Option}.value`):

```ocaml
val left_value : ('a, 'b) t -> default:'a -> 'a 
val right_value : ('a, 'b) t -> default:'b -> 'b
```
Note that Haskell doesn't qualify `fromLeft` by `Either` so I don't think we should follow their names on that one, it is confusing.

Reply:
Ah, it's just the name you're unhappy with.  I don't mind using a different name.  (I had difficulty finding the outdated comment; the only way I can find it was via an email notification.)

Reply:
I added `get_left` (now `find_left`) late in my design process. It's the only function in this API that cannot be typed with a linear arrow on the `either` input -- it discards information in some cases -- and I'm not fond of it. The reason to have it is to let people easily get a sequence-like datastructure out of their values.

I personally don't like these proposals that contain a negative encoding of the option type, they have complex type. One should be able to do the exact same thing by operating on the resulting `option` value, with a fairly simple chaining (and more! `to_seq` etc.).

If a consensus emerges that my option-returning selectors are not liked, I will remove them.